x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -444,12 +461,12 @@ boolean needsFsync() {
                         assert completedRanges != null;
                         assert completedRanges.isEmpty() == false;
 
-                        IOUtils.fsync(file, false, false); // TODO don't forget to fsync parent directory
+                        IOUtils.fsync(file, false, false);
                         success = true;
                         return completedRanges;
                     } finally {
                         if (success == false) {
-                            needsFsync.set(true);
+                            markAsNeedsFSync();
                         }
                     }
                 }
```

Comment:
If the `compareAndSet` above does not succeed, should we then perhaps fail when running tests (using an assert)?

Reply:
I think we can but the existing tests assume that fsync can be executed at any time even when fsync is not needed. Those tests should be adapted as well, maybe in a follow up?